src/chainstate/stacks/address.rs

```diff
@@ -49,46 +49,246 @@ use crate::types::chainstate::STACKS_ADDRESS_ENCODED_SIZE;
 use crate::util_lib::boot::boot_code_addr;
 
 pub trait StacksAddressExtensions {
-    fn to_bitcoin_tx_out(&self, value: u64) -> TxOut;
-    fn as_clarity_tuple(&self) -> TupleData;
     fn to_b58(self) -> String;
     fn from_bitcoin_address(addr: &BitcoinAddress) -> StacksAddress;
     fn is_boot_code_addr(&self) -> bool;
 }
 
-impl StacksAddressExtensions for StacksAddress {
-    /// is this a boot code address, if the supplied address is mainnet or testnet,
-    ///  it checks against the appropriate the boot code addr
-    fn is_boot_code_addr(&self) -> bool {
-        self == &boot_code_addr(self.is_mainnet())
+/// A PoX address as seen by the .pox and .pox-2 contracts.
+/// Used by the sortition DB and chains coordinator to extract addresses from the PoX contract to
+/// build the reward set and to validate block-commits.
+#[derive(Debug, PartialEq, PartialOrd, Ord, Clone, Hash, Eq, Serialize, Deserialize)]
+pub enum PoxAddress {
+    /// represents { version: (buff u1), hashbytes: (buff 20) }.
+    /// The address hash mode is optional because if we decode a legacy bitcoin address, we won't
+    /// be able to determine the hash mode since we can't distinguish segwit-p2sh from p2sh
+    Standard(StacksAddress, Option<AddressHashMode>),
+}
+
+impl std::fmt::Display for PoxAddress {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.to_db_string())
+    }
+}
+
+impl PoxAddress {
+    pub fn hashmode(&self) -> Option<AddressHashMode> {
+        match *self {
+            PoxAddress::Standard(_, hm) => hm.clone(),
+        }
+    }
+
+    #[cfg(any(test, feature = "testing"))]
+    pub fn version(&self) -> u8 {
+        self.hashmode()
+            .expect("FATAL: tried to load the hashmode of a PoxAddress which has none known")
+            as u8
+    }
+
+    pub fn bytes(&self) -> Vec<u8> {
+        match *self {
+            PoxAddress::Standard(addr, _) => addr.bytes.0.to_vec(),
+        }
     }
```

Comment:
Can you add rustdocs to the public functions of this struct? I think as it abstracts multiple kinds of addresses, functions like "bytes" are less self-explanatory.

Reply:
Sure, will do

Reply:
Done